### PR TITLE
Remove top-level random import to allow VM to boot faster

### PR DESCRIPTION
**Pull request by elafontaine**
Hi folks,
This is a problem I've observed on a Centos 7 machine and didn't find a bug for it yet.  I'm thinking of doing so during the day depending on the discussion that will happen here.

As the machine starts, the VM doesn't have the /dev/urandom initialised
yet.  This cause the init-local dhcp discovery to wait on /dev/urandom to reach
the meta-data server and start configuring the server, all to resolve
169.254.169.254 which shouldn't need resolution.  The impact of this is
80 sec on Redhat based OS.

As python import are cached anyway and random is only used in "rand_str"
, putting the import within the function that needs it makes sense at
the moment of writing.

https://docs.python.org/3/reference/import.html#the-module-cache

May come a time where we might want to have a test to make sure random
and other packages having similar effect never gets imported as a
top-level package in this file being used throughout cloud-init to avoid
this problem again.

### CODE REVIEW

**Comment by raharper** [comment]
> Hi folks,
> This is a problem I've observed on a Centos 7 machine and didn't find a bug for it yet. I'm thinking of doing so during the day depending on the discussion that will happen here.
> 
> As the machine starts, the VM doesn't have the /dev/urandom initialised
> yet. This cause the init-local dhcp discovery to wait on /dev/urandom to reach
> the meta-data server and start configuring the server, all to resolve
> 169.254.169.254 which shouldn't need resolution. The impact of this is
> 80 sec on Redhat based OS.

I suggest opening a bug and include the cloud-init.log so we can sort through
what timeout you're actually seeing.

**Comment by smoser** [comment]
possibly related:

 * https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1780062
 * https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1727358
 * https://www.python.org/dev/peps/pep-0524/
 * https://bugs.python.org/issue26839

**Comment by elafontaine** [comment]
Hi Folks, Greatly appreciate the conversation :).

TLDR; Good conversation and data, going to continue on launchpad :).

I do know that not importing the random in there still doesn't prevent it from being imported by another dependency.  My immediate concern (the PR) is that for Openstack spawning Centos server, the /dev/urandom block 80s due to "is_resolvable_url" in "init-local" stage.  I think the root cause is that the file being imported is called "utils".

I do like the idea proposed by @ghislainbourgeois to have some kind of "utils" for the init-local stage which would be to isolate early boot dependency vs those that require a fully running system.

@smoser really appreciate the reference :) I had read the cloud-init bug, but couldn't remember at the time of writing (I wasn't able to find it in the launchpad portal).
https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1727358

I would like to continue to discuss further about the init-local stage and init having 2 different concerns;
1- init-local is reaching the meta-data server to get the data on the machine
2- init injects entropy through the use of the "seed-random" module.

I'll be continuing the conversation over the launchpad.  If we can figure out what would be required to avoid depending on /dev/urandom to inject into /dev/urandom, I think we would benefit.


> I certainly appreciate the thought, and trimming our imports is definitely valuable, but I don't think we're going to be able to get away from importing random at startup.

@OddBloke , not sure if I was misunderstood; I don't want to optimise the import.  I want to have them happen when they actually need to happen for avoiding cases like this one (early boot stage).  I suggested a pretty ugly PR (sorry) but it was to see if I would have been breaking something and to get the conversation started :).  Sorry for the noise, but it was really beneficial to me.

BTW, do you guys have an IRC where I should actually go and connect? (if so, please let me know here or by email.

**Comment by raharper** [comment]
> BTW, do you guys have an IRC where I should actually go and connect? (if so, please let me know here or by email.

#cloud-init on FreeNode

**Comment by OddBloke** [comment]
> @OddBloke , not sure if I was misunderstood; I don't want to optimise the import. I want to have them happen when they actually need to happen for avoiding cases like this one (early boot stage).

I did understand, I think! My point was that making them happen when they actually need to happen is, unfortunately, Complicated (TM).

The imports that our dependencies perform (generally) happen when we import them. So _us_ removing `import random` from our codebase does not guarantee that `random` won't be imported by one of the other imports that we aren't removing. Indeed, considering our stdlib imports alone, we import `tempfile` and `uuid`, both of which in turn import `random`. These imports happen at the same time as the `import random`s in our codebase, so we will still pay the cost of importing `random` when they get imported even if we remove (or defer) the explicit imports.

> Sorry for the noise, but it was really beneficial to me.

No worries, the only way you can find this stuff out is by asking!

And, yes, please do join us in #cloud-init!
